File: neet_code_150_cpp/bit_manipulation/number_of_1bit.cpp
#include <cstdint>
#include <iostream>

// AC
class Solution {
    public:
        int hammingWeight(uint32_t n) {
            int sum = 0;
            for(int i = 0; i < 32; ++i) {
                if(n & (1 << i)) {
                    ++sum;
                }
            }

            return sum;
        }
};

// 模範解答
class SolutionAns {
public:
    int hammingWeight(uint32_t n) {
        int bit = 0;
        int result = 0;
        
        while (n != 0) {
            bit = n & 1;
            if (bit == 1) {
                result++;
            }
            n = n >> 1;
        }
        
        return result;
    }
};

int main(void) {
    uint32_t case_1 = 0b00000000000000000000000000001011;
    // => 3
    uint32_t case_2 = 0b00000000000000000000000010000000;
    // => 1
    uint32_t case_3 = 0b11111111111111111111111111111101;
    // => 32

    Solution s_1;
    std::cout << "case_1: " << s_1.hammingWeight(case_1) << std::endl;
    std::cout << "case_2: " << s_1.hammingWeight(case_2) << std::endl;
    std::cout << "case_3: " << s_1.hammingWeight(case_3) << std::endl;

    SolutionAns s_ans;
    std::cout << "case_1: " << s_ans.hammingWeight(case_1) << std::endl;
    std::cout << "case_2: " << s_ans.hammingWeight(case_2) << std::endl;
    std::cout << "case_3: " << s_ans.hammingWeight(case_3) << std::endl;
}

File: neet_code_150/src/bit_manipulation/number_of_1bit.rs
//AC
struct Solution {}
impl Solution {
    pub fn hamming_weight(n: u32) -> i32 {
        let mut sum = 0;
        for i in 0..32 {
            if n & (1 << i) != 0 {
                sum += 1;
            }
        }

        sum as i32
    }
}

// 模範解答
struct SolutionAns {}
impl SolutionAns {
    pub fn hamming_weight(mut n: u32) -> i32 {
        let mut count = 0;

        while n > 0 {
            n = n & (n - 1);
            count += 1;
        }

        count
    }
}

// 後で解いた別解
// AC
struct SolutionLatter;
impl SolutionLatter {
    pub fn hamming_weight(n: u32) -> i32 {
        let mut n = n;
        let mut result = 0;

        while n > 0 {
            if n & 1 != 0 {
                result += 1;
            }

            n = n >> 1;
        }

        result
    }
}

// C++の模範解答
struct SolutionAnsCpp {}
impl SolutionAnsCpp {
    pub fn hamming_weight(mut n: u32) -> i32 {
        let mut result = 0;

        while n != 0 {
            let bit = n & 1;
            if bit == 1 {
                result += 1;
            }

            n = n >> 1;
        }

        result
    }
}

fn main() {
    let case_1 = 0b00000000000000000000000000001011_u32;
    // => 3
    let case_2 = 0b00000000000000000000000010000000_u32;
    // => 1
    let case_3 = 0b11111111111111111111111111111101_u32;
    // => 31

    println!("case_1: {}", Solution::hamming_weight(case_1));
    println!("case_2: {}", Solution::hamming_weight(case_2));
    println!("case_3: {}", Solution::hamming_weight(case_3));

    println!("case_1: {}", SolutionAns::hamming_weight(case_1));
    println!("case_2: {}", SolutionAns::hamming_weight(case_2));
    println!("case_3: {}", SolutionAns::hamming_weight(case_3));

    println!("case_1: {}", SolutionAnsCpp::hamming_weight(case_1));
    println!("case_2: {}", SolutionAnsCpp::hamming_weight(case_2));
    println!("case_3: {}", SolutionAnsCpp::hamming_weight(case_3));

    println!("case_1: {}", SolutionLatter::hamming_weight(case_1));
    println!("case_2: {}", SolutionLatter::hamming_weight(case_2));
    println!("case_3: {}", SolutionLatter::hamming_weight(case_3));
}
